components/js_lang/js_lang_core.h

```diff
@@ -0,0 +1,497 @@
+    isolates (for each user) partitioned between connection? (But then the
+    question of freeing Isolates which are not in use for a long time raises).
+
+    Also we use separate JS context for each connection (and separate contexts
+    for execution of routines under different users within the same connection).
+
+    Routines which are executed under the same user in the same connection
+    share the same JS context.
+
+    TODO: Make it a class once support for KILLability/OOM handling is added.
+  */
+  struct Auth_id_context {
+    /**
+      Pointer to isolate for this specific connection and user pair.
+
+      We have to use reference counting pointer here since SQL code imposes
+      akward restrictions on Js_thd and Js_sp lifetime - Js_sp object
+      representing routine in the connection can outlive connection's Js_thd
+      object [sic!].
+    */
+    std::shared_ptr<Js_isolate> isolate_ptr;
+
+    /**
+      JS context for the specific connection and user pair.
+
+      @note Belongs to the Isolate referenced from this context.
+    */
+    v8::Global<v8::Context> context;
+
+    Auth_id_context(const std::shared_ptr<Js_isolate> &iso_ptr,
+                    v8::Local<v8::Context> &ctx)
+        : isolate_ptr(iso_ptr), context(iso_ptr->get_v8_isolate(), ctx) {}
+
+    v8::Isolate *get_isolate() const { return isolate_ptr->get_v8_isolate(); }
+
+    v8::Local<v8::Context> get_context() const {
+      return v8::Local<v8::Context>::New(isolate_ptr->get_v8_isolate(),
+                                         context);
+    }
+  };
+
+  /**
+    Get Auth_id_context representing context for the user account in the
+    current connection. Create one if necessary.
+  */
+  const Auth_id_context *get_or_create_auth_id_context(
+      const std::string &auth_id) {
+    auto i = m_auth_id_contexts.find(auth_id);
+    if (i == m_auth_id_contexts.end()) {
+      // Active user didn't run any JS in this connection before.
+      // We need to create new Isolate and Context.
+      auto isolate_ptr = Js_isolate::create();
+
+      v8::Isolate *isolate = isolate_ptr->get_v8_isolate();
+
+      // Lock the isolate and make it the current one to follow convention.
+      v8::Locker locker(isolate);
+
+      v8::Isolate::Scope isolate_scope(isolate);
+
+      v8::HandleScope handle_scope(isolate);
+
+      v8::Local<v8::Context> context = v8::Context::New(isolate);
+
+      auto r = m_auth_id_contexts.try_emplace(auth_id, isolate_ptr, context);
+
+      return &(r.first->second);
+    }
+
+    return &(i->second);
+  }
+
+  /** Checks if connection executes CREATE FUNCTION or PROCEDURE statement. */
+  bool is_create_routine_command() {
+    mysql_cstring_with_length sql_command;
+    always_ok(mysql_service_mysql_thd_attributes->get(m_thd, "sql_command",
+                                                      &sql_command));
+    return (strcmp(sql_command.str, "create_procedure") == 0 ||
```

Comment:
Similarly to what have we have discussed on Slack, it is not mandatory to check sql_command.length in this case as SQL core service returns zero-terminated string anyway. However, it might make sense to do this from performance optimization point of view. Adjusted code accordingly.